### ISSUE TITLE
Each package should know its version

### ISSUE THREAD

**Issue by certik** [hashdist/hashdist#327]
We already allow multiple versions of the same package, to be specified in the profile by the user (see e.g. [python.yaml](https://github.com/hashdist/hashstack/blob/1ce0c7a06b68db634835f3b38bae448268517ca5/pkgs/python/python.yaml)).

We should add another optional field to the yaml files, called `version`. If specified, it could be used instead of the hash to specify a package version, i.e. it would act similar to a git branch name. The `hit` tool would then show this version at various places. That way it is easy for the user to see which exact version is installed, for example of Python and other packages (which is not easy to tell from the hash).

For example, with #325, one can do:
```
$ hit show-package python@3dy43mxwqukrufgdrrlf3jng3ik7yaml
Information about the python@3dy43mxwqukr package:
Path: /local/certik/bld/python/3dy43mxwqukr
Full package hash: 3dy43mxwqukrufgdrrlf3jng3ik7yaml
List of sources:
    files:ut3iptrelcmvcl3tmyu5z67vcyza3d5d
    tar.gz:7qpecklofhkhn5uwga5mvyutvz5cgehq
List of dependencies:
    bzip2@yhn7t7sdxdfdkhyie6hg36hqvgccj7rj
    launcher@wvmh5uvayzv3bxqk5kirztzn3rckjyv7
    ncurses@viognkgus4njgat77hcakdedqccl3wjy
    openssl@m6pttxckahscjt6jdev5acigozwa6ths
    patchelf@k3rloj265ogtl4dmmmbmyt34dnffryka
    readline@5tdfuoei3z6ektgh7v7lh3ra36s32ssp
    sqlite@m5jo67qgu6zfrjydvg3fj3c5zvguflsx
    zlib@3el5ccejre7bcjqgld5gp6iym4ccd5oe
```
But this doesn't show the Python version. If the `python.yaml` had a `version` field in it, it would be saved to `artifact.json` and then it could be shown here (and it other `hit` commands).

**Comment by certik** [comment]
And one can then select various versions of packages in the stack, by adding the `version` keyword in the profile yaml next to each packae.

**Comment by dagss** [comment]
This is sort of already supported isnt it, just write "when: version == "1.2.3"" at top of spec, and pass version in the syntax above. No need to special case version wrt other parameters IMO. Any improvements we should do them to parameters in general not just version.

**Comment by certik** [comment]
I think it's mostly implemented, but I don't like the syntax, i.e. one has to add `when: version == "1.2.3"` into each spec, and also probably some default, etc.

**Comment by dagss** [comment]
But consider that in 95% of the cases you'd want to use one spec file to support many versions anyway. The below example isn't support directly either yet, but wouldn't you much rather do:

```
sources:
    - url: http://foo.com/numpy-${version}.tar.gz
      when version == 1.9.6: hash: tar.gz:23rfqwf23fd
      when version == 1.9.7: hash: tar.gz:23rfqwf23fd
      ...
build_stages:
    ...

    - when: 1.2.3 <= version <= 1.4.5  # we need a patch for there NumPy versions..
      # apply the patch

    - when: version == 1.3.x and blas.implementation == 'ATLAS'
      # apply another patch

```

rather than have a seperate spec file per version?

When we have a parameter system (which after all most package managers don't have!), I think it's much more ugly to special case "version" than keep all parameters consistent. (Special casing in the CLI tool is different.) How about you come with a more concrete syntax proposal in the spec files and we try to make it work for all parameters then...?

**Comment by certik** [comment]
I agree with everything you wrote. I would just start using it (syntax can be changed later once we gain more experience).

**Comment by dagss** [comment]
We should probably implement a version type though, so that we can have "10 > 2" and yet "1.2.3 < 1.2.4". So in practice that's a special case for version though in principle you could use that for "abi_version", "api_version" and so on too if you wanted

**Comment by certik** [comment]
We should use the same version comparison algorithm as Debian uses (see e.g. http://stackoverflow.com/questions/4957514/how-to-compare-debian-package-versions), as the version can contain various suffixes, so it is not always trivial.
